File: app/MoviesApp/app/src/main/java/com/example/moviesapp/view/screens/componentslist/ComponentsView.java
package com.example.moviesapp.view.screens.componentslist;

import androidx.recyclerview.widget.ListAdapter;

import com.example.moviesapp.data.movies.impl.MoviesRepo;
import com.example.moviesapp.data.movies.model.Movie;
import com.example.moviesapp.databinding.FragmentComponentsListBinding;
import com.example.moviesapp.view.BaseViewImpl;
import com.example.moviesapp.view.components.listitem.ListItemComponentHolder;
import com.example.ui_components.listitem.model.ListItem;

import java.util.List;
import java.util.stream.Collectors;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;

public class ComponentsView extends BaseViewImpl<FragmentComponentsListBinding> {

    private final MoviesRepo moviesRepo;
    private final CompositeDisposable compositeDisposable;
    private final ComponentsFragment.ComponentsNavigator componentsNavigator;

    ComponentsView(ComponentsFragment.ComponentsNavigator componentsNavigator) {
        moviesRepo = new MoviesRepo();
        compositeDisposable = new CompositeDisposable();
        this.componentsNavigator = componentsNavigator;
    }

    @Override
    protected void setBinding(FragmentComponentsListBinding binding) {
        ListAdapter<ListItemComponent, ListItemComponentHolder> adapter = new ComponentsAdapter();
        binding.container.setAdapter(adapter);
        compositeDisposable.add(
                moviesRepo.movies()
                        .map(this::convertToListItems)
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(adapter::submitList)
        );
        binding.executePendingBindings();
    }

    private List<ListItemComponent> convertToListItems(List<Movie> movies) {
        return movies.stream()
                .map(movie -> new ListItemComponent(
                        new ListItem(movie.getTitle()),
                        () -> componentsNavigator.openMovieDetails(movie))
                ).collect(Collectors.toList());
    }

    @Override
    protected void clearBinding() {
        compositeDisposable.clear();
    }
}

File: app/android/kotlin/MoviesApp/app/src/main/java/com/example/moviesapp/view/screens/componentslist/ComponentsView.kt
package com.example.moviesapp.view.screens.componentslist

import com.example.moviesapp.data.movies.impl.MoviesRepo
import com.example.moviesapp.data.movies.model.Movie
import com.example.moviesapp.databinding.FragmentComponentsListBinding
import com.example.moviesapp.view.BaseViewImpl
import com.example.ui_components.listitem.ListItemClickHandler
import com.example.ui_components.listitem.model.ListItem
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable

internal class ComponentsView(private val componentsNavigator: (Movie) -> Unit):
        BaseViewImpl<FragmentComponentsListBinding>() {

    private val moviesRepo = MoviesRepo()
    private val compositeDisposable = CompositeDisposable()

    override fun setBinding(binding: FragmentComponentsListBinding){
        val adapter = ComponentsAdapter()
        binding.container.adapter = adapter
        compositeDisposable.add(
                moviesRepo.movies()
                        .map(::convertToListItems)
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(adapter::submitList)
        )
        binding.executePendingBindings()
    }

    private fun convertToListItems(movies: List<Movie>) = movies.map {
        ListItemComponent(
                model = ListItem(it.title),
                handler = ListItemClickHandler { componentsNavigator(it) }
        )
    }

    override fun clearBinding() {
        compositeDisposable.clear()
    }
}